### DISCUSSION TITLE
How can I add version suffix

### DISCUSSION THREAD

**Discussion by vitoyucepi** [renovatebot/renovate#10186] [Request Help]
I want to track version of `nginx:version-alpine` docker image.
How can I add suffix to docker tag version?

```Dockerfile
FROM nginx:1.20.0-alpine

ENV TZ="UTC"

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0
```
In this `dockerfile` I use `nginx:1.12.0-alpine` image version. Also I track `NGINX_VERSION` by itself because of self-built module.
I know that this is overkill, but I want to be sure that I won't receive separate PRs because of `nginx:version` is out of sync with `nginx:version-alpine`.
In debug log I see
```json
[
  {
    "packageFile": "containers/1.20/Dockerfile",
    "deps": [
      {
        "depName": "nginx",
        "currentValue": "1.20.0-alpine",
        "replaceString": "nginx:1.20.0-alpine",
        "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
        "datasource": "docker",
        "depType": "final",
        "depIndex": 1,
        "updates": [
          {
            "bucket": "patch",
            "newVersion": "1.20.1",
            "newValue": "1.20.1-alpine",
            "newMajor": 1,
            "newMinor": 20,
            "updateType": "patch",
            "newDigest": "sha256:4c945e23f40334b2f0e8d6e5040a0ea06fc5c2e0cb355d1af0ae8ba0cdf80ea8",
            "branchName": "renovate/nginx-1.20.x"
          },
          {
            "updateType": "pin",
            "newValue": "1.20.0-alpine",
            "newDigest": "sha256:e015192ec74937149dce3aa1feb8af016b7cce3a2896246b623cfd55c14939a6",
            "branchName": "renovate/digests-pin"
          }
        ],
        "warnings": [],
        "currentVersion": "1.20.0",
        "isSingleVersion": true,
        "fixedVersion": "1.20.0-alpine"
      }
    ]
  },
  {
    "packageFile": "containers/1.20/Dockerfile",
    "deps": [
      {
        "depName": "nginx",
        "currentValue": "1.20.0",
        "datasource": "docker",
        "versioning": "semver",
        "replaceString": "# renovate: datasource=docker depName=nginx\nENV NGINX_VERSION=1.20.0\n",
        "depIndex": 0,
        "updates": [
          {
            "bucket": "patch",
            "newVersion": "1.20.1",
            "newValue": "1.20.1",
            "newMajor": 1,
            "newMinor": 20,
            "updateType": "patch",
            "branchName": "renovate/nginx-1.20.x"
          }
        ],
        "warnings": [],
        "currentVersion": "1.20.0",
        "isSingleVersion": true,
        "fixedVersion": "1.20.0"
      }
    ],
    "matchStrings": [
      "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\nENV .*?_VERSION=(?<currentValue>.*)\\n"
    ],
    "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
  }
] 
```

I use config
```json
{
  "pinDigests": true,
  "prCreation": "not-pending",
  "separateMinorPatch": true,
  "stabilityDays": 1,
  "regexManagers": [
    {
      "fileMatch": ["Dockerfile$"],
      "matchStrings": [
        "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\nENV .*?_VERSION=(?<currentValue>.*)\\n"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```

**Comment by rarkins** [comment]
You can try:

```json
{
  "packageRules": [
    {
      "matchManagers": ["regex"],
      "matchPackageNames": ["nginx"],
      "extractVersion": "^(?<version>.*)-alpine$"
    }
  ]
}
```

**Reply by vitoyucepi** [reply]
I'm not sure if this snippet works as expected.
Log looks the same
```json
{
  "depName": "nginx",
  "currentValue": "1.20.0",
  "datasource": "docker",
  "versioning": "docker",
  "replaceString": "# renovate: datasource=docker depName=nginx versioning=docker\npkgver=1.20.0",
  "depIndex": 0,
  "updates": [
    {
      "bucket": "patch",
      "newVersion": "1.20.1",
      "newValue": "1.20.1",
      "newMajor": 1,
      "newMinor": 20,
      "updateType": "patch",
      "branchName": "renovate/nginx-1.20.x"
    }
  ],
  "warnings": [],
  "currentVersion": "1.20.0",
  "isSingleVersion": true,
  "fixedVersion": "1.20.0"
}
```

**Comment by viceice** [comment]
you need to write this dockerfile:

```Dockerfile
FROM nginx:1.20.0-alpine

ENV TZ="UTC"

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0-alpine
```

**Reply by vitoyucepi** [reply]
I can modify `NGINX_VERSION` but then I have to use variable expansion everywhere `${NGINX_VERSION%-*}` so it will make `1.20.0` from `1.20.0-alpine`

**Reply by viceice** [reply]
Ok, can you explain some more detailed, what you want to achieve?

As your config posted at top should open a single pr for both version updates, as they are the same depName, so they get same branch name and are grouped automatically.

**Reply by vitoyucepi** [reply]
I am building modules for nginx. To do that I use `nginx:1.20.0-alpine`.
Every module is named `nginx-mod-name` and has version `1.20.0-r99`.
I use my own copy of https://git.alpinelinux.org/aports/tree/main/nginx/APKBUILD to build all modules at the same time.
The valid pkgver can have suffix from lists https://git.alpinelinux.org/apk-tools/tree/src/version.c#n73.
So `alpine` is not valid suffix.
To use renovate I've added comment to APKBUILD
```sh
# renovate: datasource=docker depName=nginx
pkgver=1.20.0
```
So I have to modify it like this
```sh
# renovate: datasource=docker depName=nginx
_pkgver=1.20.0-alpine
pkgver=${_pkgver%-*}
```

So after building all modules at `stage 1` in `Dockerfile`.
I have to install them at `stage 2`.
To do that I have to track version in `ENV NGINX_VERSION=1.20.0-alpine`.
So will I install them
```Dockerfile
FROM nginx:1.20.0-alpine

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0

RUN apk add --no-cache nginx-mod-name="${NGINX_VERSION}-r99"
```
Modified version will look like
```Dockerfile
FROM nginx:1.20.0-alpine

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0-alpine

RUN apk add --no-cache nginx-mod-name="${NGINX_VERSION%-*}-r99"
```

### The problem
I think that at some moment of time there will be docker tag `nginx:1.xx.1` but alpine image will be out of date `nginx:1.xx.0-alpine` for example.

**Reply by rarkins** [reply]
Can you put this into a public reproduction repo in case debugging it will help?

**Reply by vitoyucepi** [reply]
https://gitlab.com/vitoyucepi/nginx-flv-renovate
Main config https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/.gitlab/renovate.json
`Dockerfile`: https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/Dockerfile
`APKBUILD`: https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/builder/abuild/00-nginx/APKBUILD
`.gitlab-ci.yml`: https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/.gitlab-ci.yml

**Reply by rarkins** [reply]
That project is quite complex - can you describe here which parts we should be looking at specifically?

**Reply by vitoyucepi** [reply]
## `regexManagers`
https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/.gitlab/renovate/renovate-regex-managers.json
### `APKBUILD`
```json
{
  "regexManagers": [
    {
      "fileMatch": ["APKBUILD$"],
      "matchStrings": [
        "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\n_?pkgver=(?<currentValue>.*)",
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```
This part used to track versions in https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/builder/abuild/00-nginx/APKBUILD
```shell
# renovate: datasource=docker depName=nginx
pkgver=1.20.0
```
If I want to track `nginx:alpine` I have to rewrite this part of code like this
```shell
# renovate: datasource=docker depName=nginx
_pkgver=1.20.0-alpine
pkgver=${_pkgver%-*}
```
### Dockerfile
```json
{
  "regexManagers": [
    {
      "fileMatch": ["Dockerfile$"],
      "matchStrings": [
        "#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\nENV .*?_VERSION=(?<currentValue>.*)\\n"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```
This part used to track versions in https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/Dockerfile
```Dockerfile
FROM nginx:1.20.0-alpine

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0

RUN apk add --no-cache nginx-mod-name="${NGINX_VERSION}-r99"
```
If I want to track `nginx:alpine` I have to rewrite this part of code like this
```Dockerfile
FROM nginx:1.20.0-alpine

# renovate: datasource=docker depName=nginx
ENV NGINX_VERSION=1.20.0-alpine

RUN apk add --no-cache nginx-mod-name="${NGINX_VERSION%-*}-r99"
```
### Gitlab ci
```json
{
  "regexManagers": [
    {
      "fileMatch": ["\\.gitlab-ci\\.yml$"],
      "matchStrings": [
        "\\s*#\\s*renovate:\\s*datasource=(?<datasource>.*?) depName=(?<depName>.*?)( versioning=(?<versioning>.*?))?\\s*-\\s+\"?(?<currentValue>[\\d\\.]*)"
      ],
      "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}"
    }
  ]
}
```
This part used to track versions in https://gitlab.com/vitoyucepi/nginx-flv-renovate/-/blob/master/containers/1.20/.gitlab-ci.yml
```yaml
Publish container image 1.20:
  extends: .publish_docker_template
  variables:
    BUILD_TAG: "1.20"
  parallel:
    matrix:
      - PUBLISH_TAG:
          # renovate: datasource=docker depName=nginx
          - "1.20.0"
          - "1.20"
          - "stable"

```
And I **can't** rewrite this part to track `nginx-alpine` because I want to tag my image as `registry.gitlab.com/vitoyucepi/nginx-flv-renovate:1.20.0` and not `registry.gitlab.com/vitoyucepi/nginx-flv-renovate:1.20.0-alpine`

**Reply by rarkins** [reply]
So is the answer  essentially "you have to look at every part?" I was hoping you'd be able to narrow down the scope of what to look at.

**Reply by vitoyucepi** [reply]
In every file I use practically the same pattern. So it would be easier for me to add suffix to all dependencies version in renovate config.
In original question I'm asking if I can add suffix for the tracked version in renovate config somehow.
For example I can limit versions by allowed versions. But I have no idea how to extend version field value.
As far as I know I can remove parts of version value. But I can't set `compatibility` part explicitly.

As for sample of code lets use `gitlab-ci.yml`. I think only in that part I can't use `1.20.0-alpine` as version.

**Comment by viceice** [comment]
You can see both deps have same branchname, so both updates will be in same pr.
